datajoint/external.py

```diff
@@ -337,18 +340,19 @@ def delete(self, *, delete_external_files=None, limit=None, display_progress=Tru
             # delete items one by one, close to transaction-safe
             error_list = []
             for uuid, external_path in items:
-                try:
-                    count = len(self & {'hash': uuid})  # optimize
-                except Exception:
-                    pass   # if delete failed, do not remove the external file
-                else:
-                    assert count in (0, 1)
+                row = (self & {'hash': uuid}).fetch()
+                if row.size:
                     try:
-                        self._remove_external_file(external_path)
-                    except Exception as error:
-                        error_list.append((uuid, external_path, str(error)))
+                        (self & {'hash': uuid}).delete_quick()
+                    except Exception:
+                        pass   # if delete failed, do not remove the external file
                     else:
-                        (self & {'hash': uuid}).delete_quick(get_count=True)
+                        try:
+                            self._remove_external_file(external_path)
+                        except Exception as error:
+                            # adding row back into table after failed delete
+                            self.insert1(row[0])
```

Comment:
set `skip_duplicates=True` to address potential race conditions.

Reply:
I think it's okay not to re-insert if the delete fails. The procedure will return the undeleted files to the user for special handling. Without implementing an actual transaction management process, we can allow some orphaned files in the external storage and provide an additional cleanup utility.

Reply:
@dimitri-yatsenko I feel this is actually necessary. There is basically 2 primary concerns:
- **The user has only one chance to catch the errors**. Meaning that running the external delete/clean up multiple times will yield no change though it has only been removed from the external tracking table. This gives a false sense that the system has been 'cleaned-up' but it has not been removed from the actual store (this is the main complaint from the user in issue #953). We can provide an additional utility to resolve this but that would be expensive to run as it would have to 'crawl' the entire store to find objects that don't exist within the external tracking table. By simply inserting it back when there is an exception, we are allowing the error to be more visible but most importantly reproducible.
- `schema.external[store].delete()` is not currently documented that it returns a list of the errors. Users are most likely unaware of this feature and therefore aren't using it properly. @zitrosolrac Could you open an issue on this in our [datajoint-docs](https://github.com/datajoint/datajoint-docs) and reference it in #953? (Filing for now since our new team members haven't been oriented to the docs setup yet).

Reply:
```suggestion
                            self.insert1(row[0], skip_duplicates=True)
```

Reply:
@guzman-raphael Got it! I'll open an issue on this in the datajoint-docs and add the necessary reference.